### ISSUE TITLE
Create a function to get the contact name

### ISSUE THREAD

**Issue by joweste** [WhiskeySockets/Baileys#399]
**Before adding this issue, make sure you do the following to make sure this is not a duplicate:**
1. Search through the repo's previous issues
2. Read through the readme at least once
3. Search the docs for the feature you're looking for

**Just describe the feature**
I think will be a plus if we had a function like whastweb.js to get contact name from a jid.

For now, the alternative is use a store. But it doesn´t solve the problem because there are contacts such as group participants that aren´t in your contact list. 

Regards, Luiz

**Comment by FERDIZ-afk** [comment]
What do you mean by contacts that are saved on our device?

**Comment by joweste** [comment]
> What do you mean by contacts that are saved on our device?

No. I mean any contact not only saved in our device. For sample, groups participants are contacts not saved in my device and I am unable to get your names or profile pictures url

**Comment by FERDIZ-afk** [comment]
> > What do you mean by contacts that are saved on our device?
> 
> No. I mean any contact not only saved in our device. For sample, I am member from some groups and your participants are contacts not saved in my device and I am unable to get your names or profile pictures url

All the data is in the store section

**Comment by joweste** [comment]
> > > What do you mean by contacts that are saved on our device?
> > 
> > 
> > No. I mean any contact not only saved in our device. For sample, I am member from some groups and your participants are contacts not saved in my device and I am unable to get your names or profile pictures url
> 
> All the data is in the store section

Sorry, but I use baileys-bottle store and the particpants name of the groups isn´t into store. There is in the store only the contacts saved into my device.

**Comment by joweste** [comment]
Hi, 
Hi,  I am trying to get the contact name of a jid when the jid is not intro store. I search all this discord group and I dont found a solution. But,  I found a  lib cobalt(java)  in https://github.com/Auties00/Cobalt/blob/617bd77752648c994676a5834606d9378cb75033/src/main/java/it/auties/whatsapp/api/Whatsapp.java#L860 thats uses the code below.  I am not fluent in java. If someone can adapt it to javascript, I think many of us will appreciate.

```
/**
     * Queries the display name of a contact
     *
     * @param contactJid the non-null contact
     * @return a CompletableFuture
     */
    public CompletableFuture<Optional<String>> queryName(@NonNull ContactJidProvider contactJid) {
        var contact = store().findContactByJid(contactJid);
        if(contact.isPresent()){
            return CompletableFuture.completedFuture(Optional.ofNullable(contact.get().chosenName()));
        }

        var query = new MexQueryRequest(List.of(new MexQueryRequest.User(contactJid.toJid().user())), List.of("STATUS"));
        return socketHandler.sendQuery("get", "w:mex", Node.of("query", Json.writeValueAsBytes(query)))
                .thenApplyAsync(this::parseNameResponse);
    }

    private Optional<String> parseNameResponse(Node result) {
        return result.findNode("result")
                .flatMap(Node::contentAsString)
                .map(json -> Json.readValue(json, MexQueryResult.class))
                .map(MexQueryResult::data)
                .map(String::valueOf);
    } 

```

I have tried the next without success since I´m not a java guy:

```
[const number = +${p.id.replace("+", "")};
        const content = [
          {
            tag: "query",
            attrs: {},
            content: [
              {
                tag: "users",
                attrs: { user_id: number }
              },
              {
                tag: "users",
                attrs: { updates: ["STATUS"] }
              }
            ]
          }
        ];
        const info = await wbot.query({
          tag: "iq",
          attrs: {
            type: "get",
            xmlns: "w:mex"
          },
          content
        });  
]

```

**Comment by Aal-oss** [comment]
hello, can I ask how to get catalog data and then send it

**Comment by tofers** [comment]
up

**Comment by bsudy** [comment]
Hi,

I have also tried to replicate the logic and I'm getting back `xwa2_users_updates_since` info with `updates`.

Here is my code:

```javascipt
    const result = await socket.query({
      tag: 'iq',
      attrs: {
        to: S_WHATSAPP_NET,
        type: 'get',
        xmlns: 'w:mex',
      },
      content: [
        {
          tag: 'query',
          attrs: {
            // From Cobalt: https://github.com/nazarenoneto/Cobalt/blob/2340780c0378a8e769cee384a2c6a0989e4871a5/src/main/java/it/auties/whatsapp/model/request/MexQueryRequest.java#L10C15-L10C31
            query_id: '6420453474633624',
          },
          content: new TextEncoder().encode(JSON.stringify({
            variables: {
              users: jids.map((jid) => ({ user_id: jid.split('@')[0].replace('+', '') })),
              updates: ['STATUS'],
            },
          })),
        },
      ],
    });
    console.log('Results', JSON.stringify(result, undefined, 2));
    const resultNode = getBinaryNodeChild(result, 'result');
    console.log('ResultNode', JSON.stringify(resultNode, undefined, 2));
 ```
 
But I'm a bit stuck at the moment as I don't find any relevant information. It seem I can only retrieve the "about" changes from there. And event that info is not always present.

**Comment by PurpShell** [comment]
I should implement USyncQuery here!

**Comment by Akkun3704** [comment]
![Screenshot_2024-11-15-07-55-12-100_com whatsapp](https://github.com/user-attachments/assets/f54fa9d5-c58f-4f67-b00e-30bdcc7228e6)
Any update?
I can only get verified name (whatsapp business username), based on [this](https://github.com/Auties00/Cobalt/blob/master/src/main/java/it/auties/whatsapp/api/Whatsapp.java#L3077-L3080)

**Comment by purpshell** [comment]
UsyncQuery was implemented. Time to add mex support to it to get this working

**Comment by pxawtyy** [comment]
commenting just to get updated, not to really help into subject, but if someone somehow needs...

last year, i used a getName function on my WA bot that worked perfectly even using no groups:

```
sock.getName = async (jid, withoutContact = false) => {
        const id = sock.decodeJid(jid);
        withoutContact = Maria.withoutContact || withoutContact;
        let v;
    
        if (id.endsWith("@g.us")) {
            v = store.contacts[id] || {};
            if (!(v.name || v.subject)) {
                v = await sock.groupMetadata(id) || {};
            }
            return v.name || v.subject || '+' + id.replace('@g.us', '').split('@')[0];
        } else {
            v = id === '0@s.whatsapp.net' ? {
                id,
                name: 'WhatsApp'
            } : id === sock.decodeJid(Maria.user.id) ?
                sock.user :
                (store.contacts[id] || {});
    
            return (withoutContact ? '' : v.name) || v.subject || v.verifiedName || '+' + jid.replace('@s.whatsapp.net', '').split('@')[0];
        }
    }
```

but now (because of updates) this doesn't work as before... anyways, this function only works now with the store populated, as store.contacts returns:

```
{
  '55119xxxxxxxx@s.whatsapp.net': {
    id: '55119xxxxxxxx@s.whatsapp.net',
    name: '𝒯𝘢𝘩 ;-;',
  },
  '5599xxxxxxxx@s.whatsapp.net': {
    id: '5599xxxxxxxx@s.whatsapp.net',
    name: '𝐒⍺հ ᝰ',
  }, // ...
}
```

considering (in ev.on of messages.upsert):

```
store.contacts[m.sender] = { id: m.sender, name: m.pushName || m.sender };
```

**Comment by shelomito12** [comment]
Is this correct?

> **The New Reality: pushName is No Longer Available on Join**
>
> The consequence of this change in Baileys is that the temporary join name (~Name) and the pushName are no longer provided by any event when a user joins.
> 
> - The old `'messages.upsert'` notification (`messageStubType: 27`) which contained the name is gone.
> - The new `'group-participants.update'` event only provides the user's JID, not their name.
> 
> This means your bot can now only learn a new user's name in two ways:
> 
> 1. From the initial `'contacts.upsert'` sync (if the user is in your phone's address book).
> 2. When the new user sends their first message or reacts to one.

**Comment by pxawtyy** [comment]
> Is this correct?
> 
> > **The New Reality: pushName is No Longer Available on Join**
> >
> > The consequence of this change in Baileys is that the temporary join name (~Name) and the pushName are no longer provided by any event when a user joins.
> > 
> > - The old `'messages.upsert'` notification (`messageStubType: 27`) which contained the name is gone.
> > - The new `'group-participants.update'` event only provides the user's JID, not their name.
> > 
> > This means your bot can now only learn a new user's name in two ways:
> > 
> > 1. From the initial `'contacts.upsert'` sync (if the user is in your phone's address book).
> > 2. When the new user sends their first message or reacts to one.

sadly. i have a usernames.json database, with jid and name that i use to retrieve the pushname. the json gets updated once in a hour, when a new user send a message or when a existing user updates their actual name AND send a message. probably not the only way, but that's what worked for me. pushName is not available directly.

**Comment by Diegovsky** [comment]
so what is the currently blessed way of doing this?
